### PR TITLE
HMAN-25 Add Pact test

### PR DESCRIPTION
**Pull request by markdathornehmcts**
https://tools.hmcts.net/jira/browse/HMAN-25

HMAN-25 branch was created from the HMAN-22 branch.

Add contract test classes for Pact testing

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -220,6 +240,9 @@ dependencyCheck {
 }
 
 dependencyManagement {
+  imports {
+    mavenBom "org.springframework.cloud:spring-cloud-dependencies:2020.0.1"
+  }
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply build.gradle]: Removed

File: build.gradle [hunk 2]
```diff
@@ -295,6 +356,31 @@ dependencies {
     exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
   }
 
+  testImplementation group: 'io.rest-assured', name: 'rest-assured', version: versions.restAssured
+  testImplementation group: 'io.rest-assured', name: 'json-path', version: versions.restAssured
+  testImplementation group: 'io.rest-assured', name: 'xml-path', version: versions.restAssured
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply build.gradle]: Applied as suggested

File: build.gradle [hunk 3]
```diff
@@ -295,6 +356,31 @@ dependencies {
...
+  contractTestRuntimeOnly "org.junit.platform:junit-platform-commons:1.7.0"
+  contractTestRuntime("org.junit.jupiter:junit-jupiter-engine:5.7.0")
+
+  contractTestImplementation group: 'uk.gov.hmcts.reform', name: 'idam-client', version: '1.5.5'
+  contractTestImplementation group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '3.1.2'
+  contractTestImplementation group: 'org.projectlombok', name: 'lombok', version: versions.lombok
+  annotationProcessor group: 'org.projectlombok', name: 'lombok', version: versions.lombok
+  contractTestImplementation sourceSets.main.runtimeClasspath
+  contractTestImplementation sourceSets.test.runtimeClasspath
+
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply build.gradle]: Cleaned

File: src/contractTest/resources/contract-test.properties [hunk 4]
```diff
@@ -0,0 +1,4 @@
+pactbroker.host=${PACT_BROKER_URL:localhost}
+pactbroker.scheme=${PACT_BROKER_SCHEME:http}
+pactbroker.port=${PACT_BROKER_PORT:}
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/contractTest/resources/contract-test.properties]: Applied as suggested

File: src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+})
+@AutoConfigureWireMock(port = 0, stubs = "classpath:/wiremock-stubs")
+@TestPropertySource("/contract-test.properties")
+
+@ExtendWith(SpringExtension.class)
+@Provider("ccd")
+@PactBroker(scheme = "${PACT_BROKER_SCHEME:http}",
+    host = "${PACT_BROKER_URL:localhost}",
+    port = "${PACT_BROKER_PORT:80}",
+    consumerVersionSelectors = {@VersionSelector(tag = "HMAN-25")})
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java]: Applied

File: build.gradle [hunk 6]
```diff
@@ -239,6 +262,39 @@ dependencyManagement {
...
+runAndPublishConsumerPactTests.finalizedBy pactPublish
+
+task runProviderPactVerification(type:Test) {
+  logger.lifecycle("Runs provider pact Tests")
+  useJUnitPlatform()
+  testClassesDirs = sourceSets.contractTest.output.classesDirs
+  classpath = sourceSets.contractTest.runtimeClasspath
+  systemProperty 'pact.verifier.publishResults', System.getProperty('pact.verifier.publishResults')
+  systemProperty 'pact.provider.version', project.pactVersion
+  include "uk/gov/hmcts/reform/hmc/controllers/**"
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply build.gradle]: Applied

File: src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
+
+@SpringBootTest(webEnvironment = RANDOM_PORT, classes = {
+    Application.class, TestingUtil.class
+})
+@AutoConfigureWireMock(port = 0, stubs = "classpath:/wiremock-stubs")
+@TestPropertySource("/contract-test.properties")
+
+@ExtendWith(SpringExtension.class)
+@Provider("ccd")
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java]: Done

File: src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+@PactBroker(scheme = "${PACT_BROKER_SCHEME:http}",
+    host = "${PACT_BROKER_URL:localhost}",
+    port = "${PACT_BROKER_PORT:80}",
+    consumerVersionSelectors = {@VersionSelector(tag = "HMAN-25")})
+//@ContextConfiguration(classes = {ContractConfig.class, MapperConfig.class})
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java]: Gone

File: src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    @TestTemplate
+    @ExtendWith(PactVerificationSpringProvider.class)
+    public void pactVerificationTestTemplate(PactVerificationContext context) {
+        context.verifyInteraction();
+    }
+
+    @State("HMC successfully creates hearing")
+    public void createHearing() {
+        HearingRequest hearingRequest = TestingUtil.getHearingRequest();
+        doNothing().when(mockService).validateHearingRequest(hearingRequest);
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/contractTest/java/uk/gov/hmcts/reform/hmc/provider/HearingManagementProviderTest.java]: Done

**Review by RebeccaBaker — APPROVED** [review]

**Pull request merged**